src/Database/LSMTree/Internal/ByteString.hs

```diff
@@ -1,17 +1,65 @@
-{-# LANGUAGE BangPatterns #-}
-{-# LANGUAGE CPP          #-}
+{-# LANGUAGE BangPatterns  #-}
+{-# LANGUAGE CPP           #-}
+{-# LANGUAGE MagicHash     #-}
+{-# LANGUAGE UnboxedTuples #-}
 -- | @bytestring@ extras
 module Database.LSMTree.Internal.ByteString (
+    tryCheapToShort,
+    tryGetByteArray,
     shortByteStringFromTo,
     byteArrayFromTo,
 ) where
 
+import qualified Data.ByteString as BS
 import qualified Data.ByteString.Builder as BB
 import qualified Data.ByteString.Builder.Internal as BB
+import qualified Data.ByteString.Internal as BS.Internal
 import           Data.ByteString.Short (ShortByteString (SBS))
 import qualified Data.ByteString.Short.Internal as SBS
-import           Data.Primitive.ByteArray (ByteArray (..))
+import           Data.Primitive.ByteArray (ByteArray (..), emptyByteArray,
+                     sizeofByteArray)
 import           Foreign.Ptr (minusPtr, plusPtr)
+import           GHC.Exts (eqAddr#, mutableByteArrayContents#, realWorld#,
+                     unsafeFreezeByteArray#)
+import qualified GHC.ForeignPtr as Foreign
+
+-- | \( O(1) \) conversion, if possible.
+--
+-- In addition to the conditions explained for 'tryGetByteArray', the
+-- bytestring must use the full length of the underlying byte array.
+tryCheapToShort :: BS.ByteString -> Either String ShortByteString
+tryCheapToShort bs =
```

Comment:
```suggestion
unsafeTryCheapToShort :: BS.ByteString -> Either String ShortByteString
unsafeTryCheapToShort bs =
```

Reply:
I think it's actually safe. See comment below.